### PR TITLE
Microservices for Authentication and Authorization

### PR DESCRIPTION
**Pull request by PhilipAB**
**What this PR does / why we need it**:
This PR adds deployments and services for the five following new microservices:
- authn-service
- authr-service
- rbac-service
- user-service
- conversion-service

### CODE REVIEW

**Comment by ebauman** [comment]
Need to edit/remove gargantua webhook service

`Error: UPGRADE FAILED: template: hobbyfarm/templates/gargantua/service.yaml:22:22: executing "hobbyfarm/templates/gargantua/service.yaml" at <.Values.gargantua.webhook.servicePort>: nil pointer evaluating interface {}.servicePort`

**Comment by ebauman** [comment]
Need to also remove `templates/gargantua/secret.yaml` since the webhook secret is now handled in `templates/conversion-service/secret.yaml`

**Comment by ebauman** [comment]
accesscodesvc fails with the following:

`I0925 20:26:16.183564       1 main.go:67] installing/updating access code CRDs
2023-09-25T20:26:16.220066643Z F0925 20:26:16.219938       1 main.go:70] failed installing/updating access code CRDs: error getting apiextv1.CRD from k: customresourcedefinitions.apiextensions.k8s.io "accesscodes.hobbyfarm.io" is forbidden: User "system:serviceaccount:hobbyfarm-services:default" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope`

Likely need a clusterrolebinding to ensure this service can install its CRDs

**Comment by ebauman** [comment]
conversion service is giving

`F0925 21:05:14.823296       1 main.go:62] error reading ca certificate: open /etc/ssl/certs/ca.crt: no such file or directory`

The secret that mounts to this location has a field called "ca" but not "ca.crt". Thus I suspect the file is being mounted at `/etc/ssl/certs/ca` instead of `ca.crt`.

**Comment by ebauman** [comment]
Gargantua fails start

`I0925 21:12:46.861269       1 main.go:104] Starting Gargantua
2023-09-25T21:12:46.861890287Z F0925 21:12:46.861819       1 main.go:122] error reading ca certificate: open /webhook-secret/ca: no such file or directory`

**Comment by ebauman** [comment]
Rbac service starts but spits errors of

`E0925 21:21:24.522052       1 leaderelection.go:330] error retrieving resource lock hobbyfarm-services/controller-manager-rbac: configmaps "controller-manager-rbac" is forbidden: User "system:serviceaccount:hobbyfarm-services:default" cannot get resource "configmaps" in API group "" in the namespace "hobbyfarm-services"`

**Comment by ebauman** [comment]
user-service won't start, gives errors of 

`F0925 21:21:05.932833       1 main.go:117] error reading ca certificate: read /webhook-secret/ca.crt: is a directory`

**Comment by PhilipAB** [comment]
> conversion service is giving
> 
> `F0925 21:05:14.823296 1 main.go:62] error reading ca certificate: open /etc/ssl/certs/ca.crt: no such file or directory`
> 
> The secret that mounts to this location has a field called "ca" but not "ca.crt". Thus I suspect the file is being mounted at `/etc/ssl/certs/ca` instead of `ca.crt`.

Yes, correct! I forgot to update the field name to ca.crt. This should fix the error that the user service, conversion service and gargantua are not starting.

File: charts/hobbyfarm/templates/gargantua/ingress.yaml [hunk 1]
```diff
@@ -16,13 +16,62 @@ spec:
...
+            name: setting-service
+            port: 
+              number: 80
+        - pathType: Prefix
+          path: "/"
+          backend:
+            service:
+              name: gargantua
+              port: 
+                number: 80
```
Comment (ebauman) [inline charts/hobbyfarm/templates/gargantua/ingress.yaml]: needs an indent here, this fails helm's yaml parsing.

**Comment by jggoebel** [comment]
To solve the load balancing for the microservice architecture i tried the following:
Starting with:
> Accesscodeservice scaled to 2
> Authn 1 replica

Before my changes:
- Authn sticks to one of the both accesscodeservices, decided on connection start.

Change: add grpc client side load balancing
- same behaviour as before changes. connnection is sticky to one authn service

Change: change service to be headless
- same behaviour as the last change  - this one brought me the most trouble, i thought it should work this way oout of the box

Change: adding `dns:///` as a prefix for the address to enable the grpc resolver to lookup correctly
- https://github.com/grpc/grpc/blob/master/doc/naming.md
- this resolves the multiple addresses correctly
- all accesscodeservices are targeted and load is balanced


Issues to solve:

Now we scale accesscodeservice to 3 instances (without redeploying the other services like any autoscaler would do):
Behaviour:
- First 2 replicas are targeted, third one is not

Proposed solutions (to be tested yet)
- rediscovery every x seconds through custom resolver
- maxConnectionAge set to x seconds on server side, will lead clients to rediscover

Edit a few hours later:
Tested the `maxConnectionAge`. This works. Might have to figure out what a good threshold would be

**Pull request merged**